BUG: ndimage.median_filter: single element array handling

tylerjereddy opened the pull request:
* Fixes gh-23075

* This patch adds special handling for the single-element array scenario in the above ticket, along with several regression test cases that get their expected values from running the same tests against SciPy `1.14.1`, before we overhauled the `median_filter` algorithm.

[skip circle]

tylerjereddy committed:
BUG: median_filter single element array handling

* Fixes gh-23075

* This patch adds special handling for the single-element array scenario
in the above ticket, along with several regression test cases that get
their expected values from running the same tests against SciPy
`1.14.1`, before we overhauled the `median_filter` algorithm.

[skip circle]

tylerjereddy commented:
Interesting, I don't see the testing issues locally. We've seen lots of segfaulting problems with this code of course--that's why the original `return` was there.

tqbl commented on scipy/ndimage/src/_rank_filter_1d.cpp:
```diff
@@ -155,7 +155,24 @@ void _rank_filter(T *in_arr, int rank, int arr_len, int win_len, T *out_arr,
+              }
+              return;
+      }
+  }
```
Thanks @tylerjereddy. I just want to be sure we're not introducing another regression. The predicate `lim2 < 0` can be rewritten as `win_len > 2 * (arr_len + origin + 1)`, e.g. `origin=0`, `arr_len=8`, `win_len=19` would trigger it. Since the new check is only for single-element arrays, the other cases would be handled by the later parts of the code. Couldn't that potentially lead to the issues that were encountered before, like segfaults?

tylerjereddy replied:
Good point, there are two problems showing up in the CI testing at the moment.

1. `assert_allclose()` only has the `strict` kwarg in NumPy `2.x` series--that is easily fixed by using the `xp..` version of that test, which preserves the ability to check shape and dtype correctness.
2. I'm able to (stochastically) reproduce the segfault showing up in CI if use NumPy `1.26.4`--seems to be caused in `scipy/ndimage/tests/test_filters.py::test_gh_22586_crash_property`, which is the exact test we wrote to prevent this regression (so that's good, I suppose, it is doing its job). I'll see if I can do a "partial fix" for now perhaps, for example by preserving the original conditional, but special casing the length-1 arrays that are the specific source of complaint here. I suspect this will still let some corner cases cause problems, but move us a bit closer to correctness.

tylerjereddy committed:
BUG, TST: PR 23293 revisions

* The `test_gh_23075...()` tests were adjusted to use an assertion
approach that is compatible with NumPy prior to version `2.0.0` while
still preserving shape and dtype checking/stringency.

* Patch a bug described at
  https://github.com/scipy/scipy/pull/23293#discussion_r2190998466.

[skip circle]

tylerjereddy commented on scipy/ndimage/src/_rank_filter_1d.cpp:
```diff
@@ -155,7 +155,29 @@ void _rank_filter(T *in_arr, int rank, int arr_len, int win_len, T *out_arr,
+          }
+      }
+      else {
+          return;
```
@tqbl This latest patch seems to prevent the crashing caught by the regression guard, `test_gh_22586_crash_property`, by restoring the original `lim2` conditional and only special casing the scenario that is the subject of the original ticket.

That said, I suspect this is a move in the right direction, but still not the full story since the case you describe at https://github.com/scipy/scipy/pull/23293/files#r2190998466 may very well still not return the correct answer (though it should no longer segfault).

Can I convince you that this might be "ok" to delay to another issue/ticket? Probably a few permutations of cases similar to the one you describe should be tested against SciPy `1.14.1`, and we should try to see if we can find a sensible algorithm that allows us to match that previous behavior without the "lazy" `return` here.

Or, we could try to make this PR a more complete fix, though that may take a fair bit more time.

tylerjereddy replied:
Ok, I'll see what I can do. I believe my work here today is still a step in the right direction since we've reduced the number of scenarios where the premature `return` applies (i.e., the original ticket).

tqbl replied:
Thanks! Just one minor thing. `(footprint.size - 1) / 2` can be `(footprint.size - 1) // 2` to match the C++ code.

tylerjereddy committed:
BUG: PR 23293 revisions

* Add a regression test and fix for the additional
`median_filter` bug described at:
https://github.com/scipy/scipy/pull/23293/files#r2191190980

[skip circle]

tylerjereddy commented:
And now the extra shim code isn't array API compatible... pretty annoying that it is that easy to break compliance with basic arithmetic. I'll check it tomorrow.

tylerjereddy committed:
BUG: PR 23293 revisions

* Address a reviewer comment about having the Python
shims for `median_filter` better match the C++ logic they
are attempting to replicate:
https://github.com/scipy/scipy/pull/23293#discussion_r2191486256

* The patch for the above issue also fixes the array API
compatibility issue noted at:
https://github.com/scipy/scipy/pull/23293#issuecomment-3047279252

[skip circle]

tylerjereddy commented:
Applying the suggested patch from @tqbl to perform truncating division on the Python side to match the C++ guards seems to conveniently also fix the array API compatibility issue in local testing.

We'll see if the CI is happy here now.

tylerjereddy commented:
@ev-br do we have someone for reviewing `ndimage` stuff? Maybe someone on the `skimage` team if we have to? I believe they call our `ndimage` in a few of their functions for example.

tylerjereddy commented:
@lagru could you take a look maybe? I think the Wikipedia entry for Median Filter is probably sufficient background to look this over (in fact, that's most of what I used to fix it, on top of checking the previous behavior).

lagru commented:
PS: You can request a review from me. That would increase the chance that I'll get back to this.

rgommers commented on scipy/ndimage/src/_rank_filter_1d.cpp:
```diff
@@ -155,7 +155,26 @@ void _rank_filter(T *in_arr, int rank, int arr_len, int win_len, T *out_arr,
   int lim2 = arr_len - lim;
-  if (lim2 < 0) return;
+  if (lim2 < 0) {
+      if (arr_len == 1) {
```
There is no `else` clause here, which looks off. Is it impossible, missing, or did you intend for that case to fall through now to beyond line 177 in this diff, rather than hit `return;`?

tylerjereddy replied:
I've purged the `arr_len` check and the full suite still passes locally.

That said, I had intended for the last thing you mention, I really only wanted to intercept the size-1 array case and the code inside the conditional was intended for the simple scenario that was failing in the original ticket. `_filters.py` also has the size-1 exception/allowance for the fast path.  But I can't prove with any testing that we need that extra guard anymore, so while that bothers me, I've removed it.

I did double check that we still need the broader `if (lim2 < 0)` guard--there are gruesome segfaults locally without that.

tylerjereddy replied:
That's why I had it originally, but we can't prove a need for the guard with a public function test right? So this would just be a preference to have a private function behave more robustly on its own?

tylerjereddy replied:
alright, I pushed that up

rgommers commented on scipy/ndimage/_filters.py:
```diff
@@ -1989,7 +1989,8 @@ def _rank_filter(input, rank, size=None, footprint=None, output=None,
         mode = _ni_support._extend_mode_to_code(mode, is_filter=True)
-        if input.ndim == 1:
+        lim2 = input.size - ((footprint.size - 1) // 2 - origin)
+        if input.ndim == 1 and ((lim2 >= 0) or (input.size == 1)):
```
So the fix is that for the corner case of `footprint` being large compared to array size, we're skipping this branch and therefore hitting the old `_nd_image.rank_filter` C implementation rather than the new `_rank_filter_1d.rank_filter` C++ code?

I think that deserves a comment here, because future readers will wonder why the 1-D specific implementation isn't being used for `input.ndim == 1`.

tylerjereddy replied:
I've added a comment similar to that, which also refers the reader back to this PR, and suggests that it may be possible to restore the fast path more broadly with algorithmic fixes.

That said, the original fast path author has clearly indicated they don't have the bandwidth to help us maintain it. Anyway, that's clearly why we're going through this. I don't know if it has been worth it to try to keep that fast path, and what percentage of real world cases have already been sent to the slower path as a result of working around problems, so things are getting a bit murky here, but the regression tests are the one thing that makes me think we're moving forward either way, because they were obviously missing before.

rgommers reviewed (commented):
Thanks for tackling this @tylerjereddy & reviewers. 

The added test cases look good to me. A couple of questions about the implementation.

tylerjereddy committed:
MAINT: PR 23293 revisions

* Adjust array dtype specifications in `test_gh_23075()`
based on reviewer feedback.

* Simplify the `_rank_filter()` function to avoid an extra
conditional, based on reviewer feedback.

* Add a comment explaining why the 1D fast path is disabled
for some inputs to the median filter algorithm, based on
reviewer feedback.

tylerjereddy commented on scipy/ndimage/tests/test_filters.py:
```diff
@@ -3036,3 +3036,55 @@ def test_gh23046_fix(self, xp):
+    # results verified against SciPy 1.14.1, before the median_filter
+    # overhaul
+    sample_array = np.asarray(samples, dtype=np.single)
+    expected = np.asarray(expected, dtype=np.single)
```
done

tylerjereddy committed:
MAINT: PR 23293 revisions

* Make `_rank_filter()` behave a bit more robustly on its own,
with a matching comment.

rgommers reviewed (approved):
LGTM, seems good to merge and backport once CI comes back green.
